### PR TITLE
aws s3 source/sink

### PR DESCRIPTION
**Pull request by sobychacko**
This PR depends on some changes from https://github.com/spring-cloud/stream-applications/pull/59

### CODE REVIEW

**Comment by artembilan** [comment]
I think this is huge.
Let's merge that #59 first and then we can come back to this one!

**Comment by dturanski** [comment]
Should we address this comment on #sc-dataflow-support ?  Ability to connect to S3 implementations besides AWS. I don't know if it has to be addressed in a dependent library,

```Thanks all. Would be nice to have EMC ECS and min.io support out of the box. I have several use cases at my customer (german security authority like FBI) where they would like to just create flows with s3 on ECS with official starter apps deploying on TAS.```

File: functions/consumer/s3-consumer/README.adoc [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+You can use `s3Consumer` as a qualifier when injecting.
+
+## Configuration Options
+
+All configuration properties are prefixed with `s3.consumer`.
+
+For more information on the various options available, please see link:src/main/java/org/springframework/cloud/fn/consumer/s3/AwsS3ConsumerProperties.java[AwsS3ConsumerProperties].
+
+## Tests
```
Comment (dturanski) [inline functions/consumer/s3-consumer/README.adoc:20]: Examples

File: functions/consumer/s3-consumer/pom.xml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            <artifactId>spring-boot-configuration-processor</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.projectreactor</groupId>
```
Comment (dturanski) [inline functions/consumer/s3-consumer/pom.xml:47]: I don't see where this is used

**Comment by dturanski** [comment]
A lot of extra code in here. I focused on S3 related mostly

**Comment by sobychacko** [comment]
@artembilan @dturanski I needed some common code that I added into another PR (FTP) and did this PR on top of that. Once the FTP PR is merged, I will rebase this one. Sorry for the inconvenience.

File: applications/sink/s3-sink/README.adoc [hunk 3]
```diff
@@ -0,0 +1,69 @@
+//tag::ref-doc[]
+= Amazon S3 Sink
+
+This sink app supports transfer files to the Amazon S3 bucket.
+Files payloads (and directories recursively) are transferred to the `remote` directory (S3 bucket) to the `local` directory where the application is deployed.
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply applications/sink/s3-sink/README.adoc:5]: @artembilan I copied the original docs as they are. There is a local directory involved, correct?

**Comment by sobychacko** [comment]
@artembilan @dturanski PR rebased and most of the review comments have been addressed.

File: functions/common/config-common/src/main/java/org/springframework/cloud/fn/common/config/SpelExpressionConverterConfiguration.java [hunk 4]
```diff
@@ -38,7 +38,7 @@
 
 @Configuration
 @AutoConfigureAfter(name = "org.springframework.cloud.stream.config.SpelExpressionConverterConfiguration")
-@ConditionalOnMissingBean(type = "org.springframework.cloud.stream.config.SpelExpressionConverterConfiguration")
+@ConditionalOnMissingClass("org.springframework.cloud.stream.config.SpelExpressionConverterConfiguration")
```
Comment (artembilan) [inline functions/common/config-common/src/main/java/org/springframework/cloud/fn/common/config/SpelExpressionConverterConfiguration.java:41]: You said you have rebased to master.
What does this change do in this PR?

Nevertheless, I think this condition doesn't fit to the expectations.
We may have a class on CP, but exclude it from auto-configuration.
So, we will reject this config, but won't have any expected beans.

Comment (sobychacko) [reply functions/common/config-common/src/main/java/org/springframework/cloud/fn/common/config/SpelExpressionConverterConfiguration.java:41]: The change that @dtur added originally did not work. So, he had to make that change. I just introduced it in the PR after talking to him.

Comment (artembilan) [reply functions/common/config-common/src/main/java/org/springframework/cloud/fn/common/config/SpelExpressionConverterConfiguration.java:41]: Yeah... OK. Let's see how it is going to live!
Anyway I think that `SpelExpressionConverterConfiguration` has to go from SCSt...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-cloud/stream-applications/commit/cffbe2fa56217150bf59e46186835fba34969263 after fixing `AwsS3SourceTests` for the proper endpoint to stop before mocking and then starting in the beginning of the test.

**Pull request closed without merging**